qutebrowser/utils/resources.py

```diff
@@ -0,0 +1,148 @@
+#
+# You should have received a copy of the GNU General Public License
+# along with qutebrowser.  If not, see <https://www.gnu.org/licenses/>.
+
+"""Resources related utilities"""
+
+import os
+import os.path
+import io
+import re
+import sys
+import enum
+import json
+import datetime
+import traceback
+import functools
+import contextlib
+import posixpath
+import shlex
+import mimetypes
+import pathlib
+import ctypes
+import ctypes.util
+from typing import (Any, Callable, IO, Iterator, Optional,
+                    Sequence, Tuple, Type, Union,
+                    Iterable, TypeVar, TYPE_CHECKING)
+
+
+# We cannot use the stdlib version on 3.7-3.8 because we need the files() API.
+if sys.version_info >= (3, 9):
+    import importlib.resources as importlib_resources
+else:  # pragma: no cover
+    import importlib_resources
+
+import qutebrowser
+_resource_cache = {}
+
+def _resource_path(filename: str) -> pathlib.Path:
+    """Get a pathlib.Path object for a resource."""
+    assert not posixpath.isabs(filename), filename
+    assert os.path.pardir not in filename.split(posixpath.sep), filename
+
+    if hasattr(sys, 'frozen'):
+        # For PyInstaller, where we can't store resource files in a qutebrowser/ folder
+        # because the executable is already named "qutebrowser" (at least on macOS).
+        return pathlib.Path(sys.executable).parent / filename
+
+    return importlib_resources.files(qutebrowser) / filename
+
+@contextlib.contextmanager
+def _resource_keyerror_workaround() -> Iterator[None]:
+    """Re-raise KeyErrors as FileNotFoundErrors.
+
+    WORKAROUND for zipfile.Path resources raising KeyError when a file was notfound:
+    https://bugs.python.org/issue43063
+
+    Only needed for Python 3.8 and 3.9.
+    """
+    try:
+        yield
+    except KeyError as e:
+        raise FileNotFoundError(str(e))
+
+
+def _glob_resources(
+    resource_path: pathlib.Path,
+    subdir: str,
+    ext: str,
+) -> Iterable[str]:
+    """Find resources with the given extension.
+
+    Yields a resource name like "html/log.html" (as string).
+    """
+    assert '*' not in ext, ext
+    assert ext.startswith('.'), ext
+    path = resource_path / subdir
+
+    if isinstance(resource_path, pathlib.Path):
+        for full_path in path.glob(f'*{ext}'):  # . is contained in ext
+            yield full_path.relative_to(resource_path).as_posix()
+    else:  # zipfile.Path or importlib_resources compat object
+        # Unfortunately, we can't tell mypy about resource_path being of type
+        # Union[pathlib.Path, zipfile.Path] because we set "python_version = 3.6" in
+        # .mypy.ini, but the zipfiel stubs (correctly) only declare zipfile.Path with
+        # Python 3.8...
+        assert path.is_dir(), path  # type: ignore[unreachable]
+        for subpath in path.iterdir():
+            if subpath.name.endswith(ext):
+                yield posixpath.join(subdir, subpath.name)
+
+
+def preload_resources() -> None:
```

Comment:
I think it'd make sense to rename this to `preload` now that it's namespaced in `resources.py`.